src/MongoDB.Driver.Core/Core/NativeLibraryLoader/NativeLibrarySource.cs

```diff
@@ -0,0 +1,169 @@
+
+            _loader = new Lazy<INativeLibraryLoader>(() =>
+            {
+                var currentPlatform = GetCurrentPlatform();
+                if (currentPlatform.HasValue && libraryRelativePaths.TryGetValue(currentPlatform.Value, out var relativePath))
+                {
+                    return CreateLibraryLoader(currentPlatform.Value, relativePath);
+                }
+                else
+                {
+                    throw new PlatformNotSupportedException("The current platform is not supported or the library path has not been provided.");
+                }
+            }, isThreadSafe: true);
+        }
+
+        public NativeLibrarySource(IDictionary<SupportedPlatforms, string> libraryRelativePaths, string libraryName)
+            : this(libraryRelativePaths, b => libraryName)
+        {
+        }
+
+        public bool Is64BitnessPlatform
+        {
+            get
+            {
+#if NET452 || NETSTANDARD2_0
+                var is64Bit = Environment.Is64BitProcess;
+#else
+                var is64Bit = IntPtr.Size == 8;
+#endif
+                return is64Bit;
+            }
+        }
+
+        // public methods
+        public T GetFunction<T>(string name)
+        {
+            return GetFunction<T>(_loader.Value, name);
+        }
+
+        // private methods
+        private INativeLibraryLoader CreateLibraryLoader(SupportedPlatforms supportedPlatform, string relativePath)
+        {
+            var path = GetAbsolutePath(relativePath);
+            switch (supportedPlatform)
+            {
+                case SupportedPlatforms.Linux:
+                    return new Linux.NativeMethods(path);
+                case SupportedPlatforms.MacOS:
+                    return new MacOS.NativeMethods(path);
+                case SupportedPlatforms.Windows:
+                    return new Windows.NativeMethods(path);
+                default:
+                    throw new PlatformNotSupportedException("The platforms must be Windows, Linux or macOS.");
+            }
+        }
+
+        private string FindLibraryOrThrow(IList<string> basePaths, params string[] suffixPaths)
+        {
+            var candidates = new List<string>();
+            foreach (var basePath in basePaths)
+            {
+                foreach (var suffix in suffixPaths)
+                {
+                    var path = Path.Combine(basePath, suffix, _libraryName);
+                    if (File.Exists(path))
+                    {
+                        return path;
+                    }
+                    candidates.Add(path);
+                }
+            }
+
+            throw new FileNotFoundException($"Could not find {_libraryName} \n Tried: {string.Join(",", candidates)}.");
+        }
+
+        private string GetAbsolutePath(string relativePath)
+        {
+            var candidatePaths = new List<string>();
+
+            var assembly = typeof(NativeLibrarySource).GetTypeInfo().Assembly;
```

Comment:
We don't use `assembly.Location` in this line, because `xunit` copies files into temp folder before running tests. So, `assembly.Location` will contain the path to `Tmp` folder which won't have snappy libraries. At the same time, `CodeBase` property will still contain the original path. This is the reason of errors which I saw on the previous week when snappy libraries have not been found on the disk `Z`

Reply:
An alternative solution can be disabling "shadowCopy" ("shadowCopy": false) for `MongoDB.Driver.Core.Tests`.
cc: @rstam